File: backend/app/models/project.rb
class Project < ApplicationRecord
  belongs_to :user

  belongs_to :workflow_template, optional: true

  has_many :previous_runs, class_name: 'WorkflowRun', foreign_key: 'project_id'
  has_one :workflow_run

  validates :name, :description, presence: true
  # Project can be either created, pending, finshed
  # pending_approval
  validates :status,
            inclusion: {
              in: %w[created pending_workflow pending_approval approved denied],
              message: '%{value} is not a valid status'
            }
end

File: frontend/src/types/project.ts
import type User from './user';
import type { Workflowrun } from './workflowrun';

export interface Project {
  id: number;
  user?: User;
  name: string;
  status: string;
  total_cost: number;
  workflow_run?: Workflowrun;
  workflow_template?: Workflowtemplate;
  description: string;
  created_at: string;
  updated_at: string;
  url: string;
  previous_runs: Workflowrun[];
}

export interface ProjectData {
  project: Project;
}
export default Project;
